tasty/src/dotty/tools/tasty/TastyVersion.scala

```diff
@@ -0,0 +1,36 @@
+package dotty.tools.tasty
+
+case class TastyVersion private(major: Int, minor: Int, experimental: Int) {
+  def isExperimental: Boolean = experimental > 0
+
+  def nextStable: TastyVersion = copy(experimental = 0)
+
+  def minStable: TastyVersion = copy(minor = 0, experimental = 0)
+
+  def show: String = {
+    val suffix = if (isExperimental) s"-experimental-$experimental" else ""
+    s"$major.$minor$suffix"
+  }
+
+  def kind: String =
+    if (isExperimental) "experimental TASTy" else "TASTy"
+
+  def validRange: String = {
+    val min = TastyVersion(major, 0, 0)
+    val max = if (experimental == 0) this else TastyVersion(major, minor - 1, 0)
+    val extra = Option.when(experimental > 0)(this)
+    s"stable TASTy from ${min.show} to ${max.show}${extra.fold("")(e => s", or exactly ${e.show}")}"
+  }
+}
+
+object TastyVersion {
+
+  private val cache: java.util.concurrent.ConcurrentHashMap[TastyVersion, TastyVersion] =
```

Comment:
this fails the re-entrant check
```suggestion
  @scala.annotation.internal.sharable
  private val cache: java.util.concurrent.ConcurrentHashMap[TastyVersion, TastyVersion] =
```

Reply:
solved

Reply:
actually we can't use this annotation in tasty-core because it needs to compile with Scala 2

Reply:
perhaps we have to remove the re-entrant check for this project?
or we never actually published a tasty-core for scala 2, it was only to check source compatibility, so maybe we can add the scala 3 class path